File: python_pure/src/volarb_py/essvi.py
from __future__ import annotations

import math
from dataclasses import dataclass
from itertools import pairwise
from typing import Final, Literal

from volarb_py.simplex import SimplexOutcome, minimise_by_simplex
from volarb_py.svi import SviParameters, durrleman_function, total_variance
from volarb_py.svi_calibration import (
    MAXIMUM_CORRELATION,
    MINIMUM_OBSERVATIONS,
    REFERENCE_LOG_MONEYNESS_STEPS,
    SliceObservation,
    bounded_exponential,
    reference_log_moneyness,
)
from volarb_py.svi_surface import (
    DEFAULT_SURFACE_SCAN_STEPS,
    DEFAULT_TIME_STEPS_PER_INTERVAL,
    SviSurfaceSlice,
    scan_svi_surface,
)

EssviStatus = Literal[
    "converged",
    "too_few_slices",
    "too_few_observations",
    "simplex_budget_exhausted",
    "arbitrage_not_eliminated",
]

GLOBAL_PARAMETER_COUNT: Final[int] = 4
MINIMUM_ESSVI_SLICES: Final[int] = 2
BUTTERFLY_PENALTY_WEIGHT: Final[float] = 1e4
CALENDAR_PENALTY_WEIGHT: Final[float] = 1e4
PENALTY_GRID_STEPS: Final[int] = 64
MAXIMUM_POWER_LAW_EXPONENT: Final[float] = 1.0
SEED_CORRELATIONS: Final[tuple[float, ...]] = (-0.7, -0.3, 0.0)
SEED_CURVATURE_FRACTIONS: Final[tuple[float, ...]] = (0.1, 0.3, 0.6)
MAXIMUM_CURVATURE_FRACTION: Final[float] = 0.9999
MINIMUM_CURVATURE_FRACTION: Final[float] = 1e-12
DURRLEMAN_SUFFICIENT_BOUND: Final[float] = 4.0
MINIMUM_SEED_VARIANCE: Final[float] = 1e-8
MINIMUM_SEED_INCREMENT: Final[float] = 1e-10
REFERENCE_POINTS_PER_SLICE: Final[int] = REFERENCE_LOG_MONEYNESS_STEPS + 1


class InvalidEssviInputsError(ValueError):
    pass


@dataclass(frozen=True)
class EssviSliceQuotes:
    years_to_expiry: float
    observations: list[SliceObservation]


@dataclass(frozen=True)
class EssviParameters:
    atm_total_variance: list[float]
    curvature_scale: float
    power_law_exponent: float
    correlation_intercept: float
    correlation_slope: float


@dataclass(frozen=True)
class EssviCalibration:
    parameters: EssviParameters
    slices: list[SviParameters]
    objective: float
    weighted_root_mean_square_residual: float
    simplex_iterations: int
    slice_count: int
    observation_count: int
    fitted_surface: list[float]
    surface_minimum_durrleman_value: float
    surface_minimum_total_variance_time_slope: float
    status: EssviStatus


def bounded_correlation(argument: float) -> float:
    return min(max(math.tanh(argument), -MAXIMUM_CORRELATION), MAXIMUM_CORRELATION)


def atm_total_variance_from_coordinates(coordinates: list[float], slice_count: int) -> list[float]:
    levels: list[float] = []
    running = 0.0
    for index in range(slice_count):
        running += bounded_exponential(coordinates[index])
        levels.append(running)
    return levels


def power_law_exponent_from_coordinate(coordinate: float) -> float:
    return MAXIMUM_POWER_LAW_EXPONENT * 0.5 * (1.0 + math.tanh(coordinate))


def normalized_position(levels: list[float], index: int) -> float:
    span = levels[-1] - levels[0]
    if span <= 0.0:
        return 0.0
    return (levels[index] - levels[0]) / span


def bounded_power(base: float, exponent: float) -> float:
    return bounded_exponential(exponent * math.log(base))


def curvature_fraction_from_coordinate(coordinate: float) -> float:
    fraction = 0.5 * (1.0 + math.tanh(coordinate))
    return min(max(fraction, MINIMUM_CURVATURE_FRACTION), MAXIMUM_CURVATURE_FRACTION)


def maximum_curvature_scale(levels: list[float], correlations: list[float], exponent: float) -> float:
    bound = math.inf
    for level, correlation in zip(levels, correlations, strict=True):
        weight = 1.0 + abs(correlation)
        slope_bound = DURRLEMAN_SUFFICIENT_BOUND / (bounded_power(level, 1.0 - exponent) * weight)
        curvature_bound = math.sqrt(
            DURRLEMAN_SUFFICIENT_BOUND / (bounded_power(level, 1.0 - 2.0 * exponent) * weight)
        )
        bound = min(bound, slope_bound, curvature_bound)
    return bound


def parameters_from_coordinates(coordinates: list[float], slice_count: int) -> EssviParameters:
    levels = atm_total_variance_from_coordinates(coordinates, slice_count)
    exponent = power_law_exponent_from_coordinate(coordinates[slice_count + 1])
    intercept = coordinates[slice_count + 2]
    slope = coordinates[slice_count + 3]
    correlations = [
        bounded_correlation(intercept + slope * normalized_position(levels, index))
        for index in range(slice_count)
    ]
    fraction = curvature_fraction_from_coordinate(coordinates[slice_count])
    return EssviParameters(
        atm_total_variance=levels,
        curvature_scale=fraction * maximum_curvature_scale(levels, correlations, exponent),
        power_law_exponent=exponent,
        correlation_intercept=intercept,
        correlation_slope=slope,
    )


def curvature_at(parameters: EssviParameters, index: int) -> float:
    level = parameters.atm_total_variance[index]
    exponent = math.log(parameters.curvature_scale) - parameters.power_law_exponent * math.log(level)
    return bounded_exponential(exponent)


def correlation_at(parameters: EssviParameters, index: int) -> float:
    position = normalized_position(parameters.atm_total_variance, index)
    return bounded_correlation(parameters.correlation_intercept + parameters.correlation_slope * position)


def svi_slice_from_essvi(level: float, curvature: float, correlation: float) -> SviParameters:
    complement = math.sqrt(1.0 - correlation * correlation)
    return SviParameters(
        a=0.5 * level * (1.0 - correlation * correlation),
        b=0.5 * level * curvature,
        rho=correlation,
        m=-correlation / curvature,
        sigma=complement / curvature,
    )


def slices_from_parameters(parameters: EssviParameters) -> list[SviParameters]:
    return [
        svi_slice_from_essvi(
            parameters.atm_total_variance[index],
            curvature_at(parameters, index),
            correlation_at(parameters, index),
        )
        for index in range(len(parameters.atm_total_variance))
    ]


def penalty_grid(lowest_log_moneyness: float, highest_log_moneyness: float) -> list[float]:
    span = highest_log_moneyness - lowest_log_moneyness
    return [
        lowest_log_moneyness + span * index / PENALTY_GRID_STEPS for index in range(PENALTY_GRID_STEPS + 1)
    ]


def butterfly_penalty(slices: list[SviParameters], grid: list[float]) -> float:
    total = 0.0
    for parameters in slices:
        for point in grid:
            shortfall = -durrleman_function(parameters, point)
            if shortfall > 0.0:
                total += shortfall * shortfall
    return BUTTERFLY_PENALTY_WEIGHT * total


def calendar_penalty(slices: list[SviParameters], grid: list[float]) -> float:
    total = 0.0
    for earlier, later in pairwise(slices):
        for point in grid:
            shortfall = total_variance(earlier, point) - total_variance(later, point)
            if shortfall > 0.0:
                total += shortfall * shortfall
    return CALENDAR_PENALTY_WEIGHT * total


def weighted_squared_residuals(slices: list[SviParameters], quotes: list[EssviSliceQuotes]) -> float:
    total = 0.0
    for parameters, entry in zip(slices, quotes, strict=True):
        for observation in entry.observations:
            residual = total_variance(parameters, observation.log_moneyness) - observation.total_variance
            total += observation.weight * residual * residual
    return total


def calibration_objective(
    coordinates: list[float], quotes: list[EssviSliceQuotes], grid: list[float]
) -> float:
    parameters = parameters_from_coordinates(coordinates, len(quotes))
    slices = slices_from_parameters(parameters)
    return (
        weighted_squared_residuals(slices, quotes)
        + butterfly_penalty(slices, grid)
        + calendar_penalty(slices, grid)
    )


def at_the_money_variance(entry: EssviSliceQuotes) -> float:
    closest = min(entry.observations, key=lambda item: (abs(item.log_moneyness), item.log_moneyness))
    return max(closest.total_variance, MINIMUM_SEED_VARIANCE)


def level_increment_coordinates(quotes: list[EssviSliceQuotes]) -> list[float]:
    levels = [at_the_money_variance(entry) for entry in quotes]
    coordinates = [math.log(levels[0])]
    for earlier, later in pairwise(levels):
        coordinates.append(math.log(max(later - earlier, MINIMUM_SEED_INCREMENT)))
    return coordinates


def seed_coordinates(quotes: list[EssviSliceQuotes]) -> list[list[float]]:
    increments = level_increment_coordinates(quotes)
    seeds: list[list[float]] = []
    for correlation in SEED_CORRELATIONS:
        for fraction in SEED_CURVATURE_FRACTIONS:
            seeds.append(
                [
                    *increments,
                    math.atanh(2.0 * fraction - 1.0),
                    0.0,
                    math.atanh(correlation),
                    0.0,
                ]
            )
    return seeds


def fitted_surface(slices: list[SviParameters]) -> list[float]:
    points = reference_log_moneyness()
    return [total_variance(parameters, point) for parameters in slices for point in points]


def total_observation_count(quotes: list[EssviSliceQuotes]) -> int:
    return sum(len(entry.observations) for entry in quotes)


def unfitted_result(quotes: list[EssviSliceQuotes], status: EssviStatus) -> EssviCalibration:
    return EssviCalibration(
        parameters=EssviParameters([], 0.0, 0.0, 0.0, 0.0),
        slices=[],
        objective=0.0,
        weighted_root_mean_square_residual=0.0,
        simplex_iterations=0,
        slice_count=len(quotes),
        observation_count=total_observation_count(quotes),
        fitted_surface=[],
        surface_minimum_durrleman_value=0.0,
        surface_minimum_total_variance_time_slope=0.0,
        status=status,
    )


def weighted_root_mean_square_residual(slices: list[SviParameters], quotes: list[EssviSliceQuotes]) -> float:
    total_weight = sum(observation.weight for entry in quotes for observation in entry.observations)
    if total_weight <= 0.0:
        return 0.0
    return math.sqrt(weighted_squared_residuals(slices, quotes) / total_weight)


def validate_quotes(quotes: list[EssviSliceQuotes]) -> None:
    for entry in quotes:
        if entry.years_to_expiry <= 0.0:
            raise InvalidEssviInputsError(f"years_to_expiry must be positive, got {entry.years_to_expiry}")
    for earlier, later in pairwise(quotes):
        if later.years_to_expiry <= earlier.years_to_expiry:
            raise InvalidEssviInputsError("slices must be strictly increasing in years_to_expiry")


def calibration_status(settled: bool, surface_status: str) -> EssviStatus:
    if surface_status != "arbitrage_free_on_grid":
        return "arbitrage_not_eliminated"
    if not settled:
        return "simplex_budget_exhausted"
    return "converged"


def calibrate_essvi_surface(
    quotes: list[EssviSliceQuotes], lowest_log_moneyness: float, highest_log_moneyness: float
) -> EssviCalibration:
    if len(quotes) < MINIMUM_ESSVI_SLICES:
        return unfitted_result(quotes, "too_few_slices")
    validate_quotes(quotes)
    if any(len(entry.observations) < MINIMUM_OBSERVATIONS for entry in quotes):
        return unfitted_result(quotes, "too_few_observations")
    if not highest_log_moneyness > lowest_log_moneyness:
        raise InvalidEssviInputsError("the penalty range must be non-empty and increasing")

    grid = penalty_grid(lowest_log_moneyness, highest_log_moneyness)
    best_outcome: SimplexOutcome | None = None
    total_iterations = 0
    for seed in seed_coordinates(quotes):
        outcome = minimise_by_simplex(
            lambda coordinates: calibration_objective(coordinates, quotes, grid), seed
        )
        total_iterations += outcome.iterations
        if best_outcome is None or outcome.value < best_outcome.value:
            best_outcome = outcome

    assert best_outcome is not None
    parameters = parameters_from_coordinates(best_outcome.coordinates, len(quotes))
    slices = slices_from_parameters(parameters)
    scan = scan_svi_surface(
        [
            SviSurfaceSlice(entry.years_to_expiry, fitted)
            for entry, fitted in zip(quotes, slices, strict=True)
        ],
        lowest_log_moneyness,
        highest_log_moneyness,
        DEFAULT_SURFACE_SCAN_STEPS,
        DEFAULT_TIME_STEPS_PER_INTERVAL,
    )
    return EssviCalibration(
        parameters=parameters,
        slices=slices,
        objective=best_outcome.value,
        weighted_root_mean_square_residual=weighted_root_mean_square_residual(slices, quotes),
        simplex_iterations=total_iterations,
        slice_count=len(quotes),
        observation_count=total_observation_count(quotes),
        fitted_surface=fitted_surface(slices),
        surface_minimum_durrleman_value=scan.minimum_durrleman_value,
        surface_minimum_total_variance_time_slope=scan.minimum_total_variance_time_slope,
        status=calibration_status(best_outcome.settled, scan.status),
    )

File: cpp_core/src/essvi.cpp
#include "volarb/essvi.hpp"

#include "volarb/simplex.hpp"
#include "volarb/svi_surface.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <optional>

namespace volarb {

namespace {

const std::vector<double> essvi_seed_correlations = {-0.7, -0.3, 0.0};
const std::vector<double> essvi_seed_curvature_fractions = {0.1, 0.3, 0.6};

double bounded_correlation(double argument) {
    return std::min(std::max(std::tanh(argument), -maximum_correlation), maximum_correlation);
}

std::vector<double> atm_total_variance_from_coordinates(const std::vector<double>& coordinates,
                                                        std::size_t slice_count) {
    std::vector<double> levels;
    levels.reserve(slice_count);
    double running = 0.0;
    for (std::size_t index = 0; index < slice_count; ++index) {
        running += bounded_exponential(coordinates[index]);
        levels.push_back(running);
    }
    return levels;
}

double power_law_exponent_from_coordinate(double coordinate) {
    return maximum_power_law_exponent * 0.5 * (1.0 + std::tanh(coordinate));
}

double normalized_position(const std::vector<double>& levels, std::size_t index) {
    const double span = levels.back() - levels.front();
    if (span <= 0.0) {
        return 0.0;
    }
    return (levels[index] - levels.front()) / span;
}

double bounded_power(double base, double exponent) {
    return bounded_exponential(exponent * std::log(base));
}

double curvature_fraction_from_coordinate(double coordinate) {
    const double fraction = 0.5 * (1.0 + std::tanh(coordinate));
    return std::min(std::max(fraction, minimum_curvature_fraction), maximum_curvature_fraction);
}

std::vector<double> penalty_grid(double lowest_log_moneyness, double highest_log_moneyness) {
    const double span = highest_log_moneyness - lowest_log_moneyness;
    std::vector<double> grid;
    grid.reserve(static_cast<std::size_t>(essvi_penalty_grid_steps + 1));
    for (int index = 0; index <= essvi_penalty_grid_steps; ++index) {
        grid.push_back(lowest_log_moneyness +
                       span * static_cast<double>(index) / static_cast<double>(essvi_penalty_grid_steps));
    }
    return grid;
}

double butterfly_penalty(const std::vector<SviParameters>& slices, const std::vector<double>& grid) {
    double total = 0.0;
    for (const SviParameters& parameters : slices) {
        for (const double point : grid) {
            const double shortfall = -durrleman_function(parameters, point);
            if (shortfall > 0.0) {
                total += shortfall * shortfall;
            }
        }
    }
    return essvi_butterfly_penalty_weight * total;
}

double calendar_penalty(const std::vector<SviParameters>& slices, const std::vector<double>& grid) {
    double total = 0.0;
    for (std::size_t index = 1; index < slices.size(); ++index) {
        for (const double point : grid) {
            const double shortfall =
                total_variance(slices[index - 1], point) - total_variance(slices[index], point);
            if (shortfall > 0.0) {
                total += shortfall * shortfall;
            }
        }
    }
    return essvi_calendar_penalty_weight * total;
}

double weighted_squared_residuals(const std::vector<SviParameters>& slices,
                                  const std::vector<EssviSliceQuotes>& quotes) {
    double total = 0.0;
    for (std::size_t index = 0; index < quotes.size(); ++index) {
        for (const SliceObservation& observation : quotes[index].observations) {
            const double residual =
                total_variance(slices[index], observation.log_moneyness) - observation.total_variance;
            total += observation.weight * residual * residual;
        }
    }
    return total;
}

double calibration_objective(const std::vector<double>& coordinates,
                             const std::vector<EssviSliceQuotes>& quotes,
                             const std::vector<double>& grid) {
    const EssviParameters parameters = parameters_from_coordinates(coordinates, quotes.size());
    const std::vector<SviParameters> slices = slices_from_parameters(parameters);
    return weighted_squared_residuals(slices, quotes) + butterfly_penalty(slices, grid) +
           calendar_penalty(slices, grid);
}

double at_the_money_variance(const EssviSliceQuotes& entry) {
    const SliceObservation* closest = &entry.observations.front();
    for (const SliceObservation& observation : entry.observations) {
        const double distance = std::abs(observation.log_moneyness);
        const double best = std::abs(closest->log_moneyness);
        if (distance < best || (distance == best && observation.log_moneyness < closest->log_moneyness)) {
            closest = &observation;
        }
    }
    return std::max(closest->total_variance, minimum_seed_variance);
}

std::vector<double> level_increment_coordinates(const std::vector<EssviSliceQuotes>& quotes) {
    std::vector<double> levels;
    levels.reserve(quotes.size());
    for (const EssviSliceQuotes& entry : quotes) {
        levels.push_back(at_the_money_variance(entry));
    }
    std::vector<double> coordinates{std::log(levels.front())};
    for (std::size_t index = 1; index < levels.size(); ++index) {
        coordinates.push_back(std::log(std::max(levels[index] - levels[index - 1], minimum_seed_increment)));
    }
    return coordinates;
}

std::vector<Coordinates> seed_coordinates(const std::vector<EssviSliceQuotes>& quotes) {
    const std::vector<double> increments = level_increment_coordinates(quotes);
    std::vector<Coordinates> seeds;
    for (const double correlation : essvi_seed_correlations) {
        for (const double fraction : essvi_seed_curvature_fractions) {
            Coordinates seed = increments;
            seed.push_back(std::atanh(2.0 * fraction - 1.0));
            seed.push_back(0.0);
            seed.push_back(std::atanh(correlation));
            seed.push_back(0.0);
            seeds.push_back(seed);
        }
    }
    return seeds;
}

std::vector<double> fitted_surface(const std::vector<SviParameters>& slices) {
    const std::vector<double> points = reference_log_moneyness();
    std::vector<double> surface;
    surface.reserve(slices.size() * points.size());
    for (const SviParameters& parameters : slices) {
        for (const double point : points) {
            surface.push_back(total_variance(parameters, point));
        }
    }
    return surface;
}

int total_observation_count(const std::vector<EssviSliceQuotes>& quotes) {
    std::size_t total = 0;
    for (const EssviSliceQuotes& entry : quotes) {
        total += entry.observations.size();
    }
    return static_cast<int>(total);
}

EssviCalibration unfitted_result(const std::vector<EssviSliceQuotes>& quotes, EssviStatus status) {
    return EssviCalibration{EssviParameters{{}, 0.0, 0.0, 0.0, 0.0},
                            {},
                            0.0,
                            0.0,
                            0,
                            static_cast<int>(quotes.size()),
                            total_observation_count(quotes),
                            {},
                            0.0,
                            0.0,
                            status};
}

double weighted_root_mean_square_residual(const std::vector<SviParameters>& slices,
                                          const std::vector<EssviSliceQuotes>& quotes) {
    double total_weight = 0.0;
    for (const EssviSliceQuotes& entry : quotes) {
        for (const SliceObservation& observation : entry.observations) {
            total_weight += observation.weight;
        }
    }
    if (total_weight <= 0.0) {
        return 0.0;
    }
    return std::sqrt(weighted_squared_residuals(slices, quotes) / total_weight);
}

void validate_quotes(const std::vector<EssviSliceQuotes>& quotes) {
    for (const EssviSliceQuotes& entry : quotes) {
        if (entry.years_to_expiry <= 0.0) {
            throw InvalidEssviInputsError("years_to_expiry must be positive, got " +
                                          std::to_string(entry.years_to_expiry));
        }
    }
    for (std::size_t index = 1; index < quotes.size(); ++index) {
        if (quotes[index].years_to_expiry <= quotes[index - 1].years_to_expiry) {
            throw InvalidEssviInputsError("slices must be strictly increasing in years_to_expiry");
        }
    }
}

EssviStatus calibration_status(bool settled, SviSurfaceStatus surface_status) {
    if (surface_status != SviSurfaceStatus::ArbitrageFreeOnGrid) {
        return EssviStatus::ArbitrageNotEliminated;
    }
    if (!settled) {
        return EssviStatus::SimplexBudgetExhausted;
    }
    return EssviStatus::Converged;
}

}

std::string name_of_essvi_status(EssviStatus status) {
    switch (status) {
    case EssviStatus::Converged:
        return "converged";
    case EssviStatus::TooFewSlices:
        return "too_few_slices";
    case EssviStatus::TooFewObservations:
        return "too_few_observations";
    case EssviStatus::SimplexBudgetExhausted:
        return "simplex_budget_exhausted";
    case EssviStatus::ArbitrageNotEliminated:
        return "arbitrage_not_eliminated";
    }
    return "arbitrage_not_eliminated";
}

double maximum_curvature_scale(const std::vector<double>& levels,
                               const std::vector<double>& correlations, double exponent) {
    double bound = std::numeric_limits<double>::infinity();
    for (std::size_t index = 0; index < levels.size(); ++index) {
        const double weight = 1.0 + std::abs(correlations[index]);
        const double slope_bound =
            durrleman_sufficient_bound / (bounded_power(levels[index], 1.0 - exponent) * weight);
        const double curvature_bound = std::sqrt(
            durrleman_sufficient_bound / (bounded_power(levels[index], 1.0 - 2.0 * exponent) * weight));
        bound = std::min(bound, std::min(slope_bound, curvature_bound));
    }
    return bound;
}

EssviParameters parameters_from_coordinates(const std::vector<double>& coordinates,
                                            std::size_t slice_count) {
    const std::vector<double> levels = atm_total_variance_from_coordinates(coordinates, slice_count);
    const double exponent = power_law_exponent_from_coordinate(coordinates[slice_count + 1]);
    const double intercept = coordinates[slice_count + 2];
    const double slope = coordinates[slice_count + 3];
    std::vector<double> correlations;
    correlations.reserve(slice_count);
    for (std::size_t index = 0; index < slice_count; ++index) {
        correlations.push_back(bounded_correlation(intercept + slope * normalized_position(levels, index)));
    }
    const double fraction = curvature_fraction_from_coordinate(coordinates[slice_count]);
    return EssviParameters{levels, fraction * maximum_curvature_scale(levels, correlations, exponent),
                           exponent, intercept, slope};
}

double curvature_at(const EssviParameters& parameters, std::size_t index) {
    const double level = parameters.atm_total_variance[index];
    const double exponent =
        std::log(parameters.curvature_scale) - parameters.power_law_exponent * std::log(level);
    return bounded_exponential(exponent);
}

double correlation_at(const EssviParameters& parameters, std::size_t index) {
    const double position = normalized_position(parameters.atm_total_variance, index);
    return bounded_correlation(parameters.correlation_intercept +
                               parameters.correlation_slope * position);
}

SviParameters svi_slice_from_essvi(double level, double curvature, double correlation) {
    const double complement = std::sqrt(1.0 - correlation * correlation);
    return SviParameters{0.5 * level * (1.0 - correlation * correlation), 0.5 * level * curvature,
                         correlation, -correlation / curvature, complement / curvature};
}

std::vector<SviParameters> slices_from_parameters(const EssviParameters& parameters) {
    std::vector<SviParameters> slices;
    slices.reserve(parameters.atm_total_variance.size());
    for (std::size_t index = 0; index < parameters.atm_total_variance.size(); ++index) {
        slices.push_back(svi_slice_from_essvi(parameters.atm_total_variance[index],
                                              curvature_at(parameters, index),
                                              correlation_at(parameters, index)));
    }
    return slices;
}

EssviCalibration calibrate_essvi_surface(const std::vector<EssviSliceQuotes>& quotes,
                                         double lowest_log_moneyness, double highest_log_moneyness) {
    if (quotes.size() < minimum_essvi_slices) {
        return unfitted_result(quotes, EssviStatus::TooFewSlices);
    }
    validate_quotes(quotes);
    for (const EssviSliceQuotes& entry : quotes) {
        if (entry.observations.size() < static_cast<std::size_t>(minimum_observations)) {
            return unfitted_result(quotes, EssviStatus::TooFewObservations);
        }
    }
    if (!(highest_log_moneyness > lowest_log_moneyness)) {
        throw InvalidEssviInputsError("the penalty range must be non-empty and increasing");
    }

    const std::vector<double> grid = penalty_grid(lowest_log_moneyness, highest_log_moneyness);
    const Objective evaluate = [&](const Coordinates& coordinates) {
        return calibration_objective(coordinates, quotes, grid);
    };

    std::optional<SimplexOutcome> best_outcome;
    int total_iterations = 0;
    for (const Coordinates& seed : seed_coordinates(quotes)) {
        const SimplexOutcome outcome = minimise_by_simplex(seed, evaluate);
        total_iterations += outcome.iterations;
        if (!best_outcome.has_value() || outcome.value < best_outcome->value) {
            best_outcome = outcome;
        }
    }

    const EssviParameters parameters = parameters_from_coordinates(best_outcome->coordinates, quotes.size());
    const std::vector<SviParameters> slices = slices_from_parameters(parameters);

    std::vector<SviSurfaceSlice> surface;
    surface.reserve(quotes.size());
    for (std::size_t index = 0; index < quotes.size(); ++index) {
        surface.push_back(SviSurfaceSlice{quotes[index].years_to_expiry, slices[index]});
    }
    const SviSurfaceScan scan =
        scan_svi_surface(surface, lowest_log_moneyness, highest_log_moneyness,
                         default_surface_scan_steps, default_time_steps_per_interval);

    return EssviCalibration{parameters,
                            slices,
                            best_outcome->value,
                            weighted_root_mean_square_residual(slices, quotes),
                            total_iterations,
                            static_cast<int>(quotes.size()),
                            total_observation_count(quotes),
                            fitted_surface(slices),
                            scan.minimum_durrleman_value,
                            scan.minimum_total_variance_time_slope,
                            calibration_status(best_outcome->settled, scan.status)};
}

}
